### PR TITLE
My Site Dashboard - Phase 2 - Draft and Scheduled Post Cards

### PR DESCRIPTION
**Pull request by ashiagr**
Parent #15200

This PR adds draft and scheduled post cards using mocked data.

  light_mode_post_cards

To test:

1. Go to `My Site` -> `Me` -> `App Settings` -> `Test Feature Configuration`.
2. Turn the `MySiteDashboardPhase2FeatureConfig` feature flag `ON`.
3. Relaunch the app.
4. Note that on the `My Site` tab
    - Draft and Scheduled post cards are present initially.
    - Card visibility based on draft/ scheduled posts availability is handled (can be tested by updating mocked data).
    - Null title/excerpt/ featured image is handled (can be tested by updating mocked data).
    - RTL is supported.

Notes:
1. The PR includes basic styling and doesn't require a design review yet. There's a separate fine-tune UI task which will include a design review for all post cards. 
2. When no title is present for a post, "Untitled" is shown in the title. 
3. As discussed, single post item with two-line excerpt is not supported due to UI complexity. If needed, we'll get back to it in a future PR. 

## Regression Notes
1. Potential unintended areas of impact
To be consistent with post cards, currently, no text style changes are applied to other cards (e.g.  Quick Links or Quick Start Cards) as post cards are behind a feature flag. Will discuss and apply them in the fine-tune UI task.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None

3. What automated tests I added (or what prevented me from doing so)
Added unit tests for new post cards.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt]: Minor (🔍): Since this `build(...)` function will return an implicit `MutableList<PostCard>` due to the `mutableListOf<PostCard>()` construction, consider making the return type explicit and make the return type `List<PostCard>` so that it no longer becomes available to the client as a mutable list.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:45]: Suggestion (💡): I was playing around with this, the `PostCardDraftOrScheduled` data class and its inner `PostItem` and something was itching me in terms of whether this can become a little more separated and/or simpler.

I then went and started splitting the `PostCardDraftOrScheduled` into a `DraftPostCard` and a `ScheduledPostCard`, which guide me through some refactorings. Long story short, I managed to end up with the following solution. Please consider that, along with it's pros and cons, but feel free to ignore it as well:

`PostItemType` (renamed from `PostCardType`)
```
enum class PostItemType(val id: Int) {
    DRAFT(0),
    SCHEDULED(1)
}
```

`PostCard`
```
        sealed class PostCard(
            open val title: UiString,
            open val postItems: List<PostItem>
        ) : Card(POST_CARD) {
            data class DraftPostCard(
                override val title: UiString,
                override val postItems: List<PostItem>
            ) : PostCard(title, postItems)

            data class ScheduledPostCard(
                override val title: UiString,
                override val postItems: List<PostItem>
            ) : PostCard(title, postItems)

            data class PostItem(
                val type: PostItemType,
                val title: UiString,
                val excerpt: UiString?,
                val featuredImageUrl: String?
            )
        }
```

`PostCardBuilder`
```
    private fun List<Post>.createDraftPostsCard() = DraftPostCard(
            title = UiStringRes(R.string.my_site_post_card_draft_title),
            postItems = mapToDraftPostItems()
    )

    private fun List<Post>.createScheduledPostsCard() = ScheduledPostCard(
            title = UiStringRes(R.string.my_site_post_card_scheduled_title),
            postItems = mapToScheduledPostItems()
    )

    private fun List<Post>.mapToDraftPostItems() = map { post ->
        PostCard.PostItem(
                type = PostItemType.DRAFT,
                title = post.title?.let { UiStringText(it) } ?: UiStringRes(R.string.my_site_untitled_post),
                excerpt = post.excerpt?.let { UiStringText(it) },
                featuredImageUrl = post.featuredImageUrl
        )
    }

    private fun List<Post>.mapToScheduledPostItems() = map { post ->
        PostCard.PostItem(
                type = PostItemType.SCHEDULED,
                title = post.title?.let { UiStringText(it) } ?: UiStringRes(R.string.my_site_untitled_post),
                excerpt = UiStringText("Today at 1:04 PM"),
                featuredImageUrl = post.featuredImageUrl
        )
    }
```

`PostCardViewHolder`
```
    fun bind(card: PostCard) = with(binding) {
        mySiteToolbar.update(card)
        (postItems.adapter as PostItemsAdapter).update(card.postItems)
    }

    private fun MySiteCardToolbarBinding.update(card: PostCard) {
        mySiteCardToolbarTitle.text = uiHelpers.getTextOfUiString(
                itemView.context,
                card.title
        )
    }
```

`PostItemViewHolder`
```
    fun bind(postItem: PostItem) = with(binding) {
        uiHelpers.setTextOrHide(title, postItem.title)
        uiHelpers.setTextOrHide(excerpt, postItem.excerpt)
        postItem.featuredImageUrl?.let { bindFeatureImage(it) }
        bindIconTime(postItem.type)
    }

    private fun PostItemBinding.bindFeatureImage(featuredImageUrl: String) {
        imageManager.loadImageWithCorners(
                featuredImage,
                PHOTO_ROUNDED_CORNERS,
                featuredImageUrl,
                uiHelpers.getPxOfUiDimen(
                        WordPress.getContext(),
                        UIDimenRes(dimen.my_site_post_item_image_corner_radius)
                )
        )
        featuredImage.setVisible(true)
    }

    private fun PostItemBinding.bindIconTime(type: PostItemType) {
        if (type == SCHEDULED) iconTime.setVisible(true)
    }
```

`post_item.xml`
```
    <ImageView
        android:id="@+id/featured_image"
        android:layout_width="@dimen/my_site_post_item_image_size"
        android:layout_height="@dimen/my_site_post_item_image_size"
        android:contentDescription="@string/featured_image_desc"
        android:scaleType="centerCrop"
        android:visibility="gone"
        app:layout_constraintBottom_toBottomOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintTop_toTopOf="@id/title"
        tools:srcCompat="@tools:sample/avatars"
        tools:visibility="visible" />
```

PS.1: I renamed the enum to `PostItemType` because the `type` is now used within the `PostItem` data class and as such it felt that changing that was appropriate.
PS.2: I removed the `isFeaturedImageVisible` from the `PostItem` since it felt as being an unnecessary duplication. However, I am not sure if you are thinking of it from another perspective. Instead, I updated the `featured_image` image view to a default visibility of `gone`, just like you did with the `icon_time` image view. If we expect that most of the post will have a feature image, we might then want to have the default visibility to `visible`, and only set the `featuredImage.setVisible(false)` when that is not presented, as such this will become the opposite of the above solution.
PS.3: I am very unsure about why the `featuredImageCornerRadius` was part of the `PostItem` and thus I took the liberty of removing it and (kind of) hardcoding it within the `PostItemViewHolder` directly. Let me know if I am missing something.
PS.4: The `PostCardBuilderTest` will need to be updated as well, but nothing too dramatic here...

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:34]: Question (❓): This will be updated in a separate PR, right?

File: WordPress/src/main/res/raw/mocked_posts_data.json
Comment (ParaskP7) [inline WordPress/src/main/res/raw/mocked_posts_data.json:10]: 1. Suggestion (💡): Let's have one `featured_image_url` as `null` for the drafts (this first draft) and one more for the scheduled (say the second scheduled), then we will win this extra visual combination during testing.
2. Suggestion (💡): Let's also have one `title` as `null` for, say, the first scheduled and win this extra visual combination during testing as well.

**Review by ParaskP7 — APPROVED** [review]
👋 @ashiagr !

I have reviewed and tested this PR as per the instructions, everything works as expected, great job! 🌟 🌟  

I have left one question (❓), some minor (🔍) comments and few suggestions (💡) for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:34]: Right, it will be updated in a separate PR based on the modified date format in the API.

File: WordPress/src/main/res/raw/mocked_posts_data.json
Comment (ashiagr) [reply WordPress/src/main/res/raw/mocked_posts_data.json:10]: In 59eeccd

I also included a long title and updated the screenshot in the PR description.

Comment (ParaskP7) [reply WordPress/src/main/res/raw/mocked_posts_data.json:10]: Great, thank Youuuuuuuuu!

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:34]: Great, thanks for letting me know Ashita! 🙏

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt]: Great, I am already seeing it as done through 34fb96ee037f454d7eab1e2ff6dc7b4e8f8061a0, thank you Ashita! 🙏

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:45]: Thank you @ParaskP7 for all the suggestions here and discussions in our call. 🙌 

As discussed, both the existing and suggested approaches are fine and each one has its pros and cons considering:
- What do we consider as a post item (1:1 mapping with an existing post?) and whether other cards need a post items list which will drive the decision if the parent `PostCard` should include `postItems`.
- Nesting level in the `PostCard` sealed class when we add other cards vs 1:1 mapping of data classes with their cards.
- Code duplication in layouts (draft and scheduled cards have more-or-less same UI structure) vs nullability. 

With all that in mind, we decided to revisit the approach after all the post cards are implemented and merged (added it as a sub-task in the parent issue). 

> PS.1: I renamed the enum to PostItemType because the type is now used within the PostItem data class and as such it felt that changing that was appropriate.

As discussed, we will reconsider it later based on what approach we take up.

> PS.2: I removed the isFeaturedImageVisible from the PostItem since it felt as being an unnecessary duplication. However, I am not sure if you are thinking of it from another perspective. Instead, I updated the featured_image image view to a default visibility of gone, just like you did with the icon_time image view. If we expect that most of the post will have a feature image, we might then want to have the default visibility to visible, and only set the featuredImage.setVisible(false) when that is not presented, as such this will become the opposite of the above solution.

The field `isFeaturedImageVisible` was added mainly to avoid any logic in the VH (`postItem.featuredImageUrl != null` ) and to directly use it in the tests to know if the image is visible or not. To avoid duplication, I've accepted your suggestion in 96ca924.

> PS.3: I am very unsure about why the featuredImageCornerRadius was part of the PostItem and thus I took the liberty of removing it and (kind of) hardcoding it within the PostItemViewHolder directly. Let me know if I am missing something.

Suggestion accepted in a6ea9bc.

> PS.4: The PostCardBuilderTest will need to be updated as well, but nothing too dramatic here...

N/A as we decided not to change the approach right now.

Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/post/PostCardBuilder.kt:45]: 👋 @ashiagr !

Thank you so much for this nice summary of what was discussed and for all the little improvements that you made based on my `PS`, much appreciated. 🙏

**Comment by ParaskP7** [comment]
👋 @ashiagr !

Everything LGTM, please feel free to merge this PR when you need to.

**Pull request merged**
